salt/minion.py

```diff
@@ -578,12 +581,19 @@ def eval_master(self,
         # happy with the first one that allows us to connect
         if isinstance(opts['master'], list):
             conn = False
-            # shuffle the masters and then loop through them
+            last_exc = None
+            opts['master_uri_list'] = []
             opts['local_masters'] = copy.copy(opts['master'])
+
+            # shuffle the masters and then loop through them
             if opts['random_master']:
-                shuffle(opts['local_masters'])
-            last_exc = None
-            opts['master_uri_list'] = list()
+                # master_failback is only used when master_type is set to failover
+                if opts['master_type'] == 'failover' and opts['master_failback']:
+                    secondary_masters = opts['local_masters'][1:]
+                    shuffle(secondary_masters)
+                    opts['local_masters'][1:] = secondary_masters
```

Comment:
Maybe we should clarify so that only secondary masters are shuffled?

Reply:
Do you mean the [master_failback option](https://docs.saltstack.com/en/latest/ref/configuration/minion.html#master-failback) docs? It seems implied that the first master wouldn't be shuffled, but I can see how that might be confusing.

Reply:
@rallytime yeah, that part. And they need to know that since `secondary_masters` are attached to the `opts['local_masters'][1:]` has implications too.